basis/http/client/client.factor

```diff
@@ -111,15 +142,73 @@ SYMBOL: redirects
     "transfer-encoding" header "chunked" =
     [ read-chunked ] [ each-block ] if ; inline
 
+: request-socket-endpoints ( request -- physical logical )
+    [ proxy-url>> ] [ url>> ] bi [ or ] keep ;
+
 : <request-socket> ( -- stream )
-    request get url>> url-addr ascii <client> drop
+    request get request-socket-endpoints [ url-addr ] bi@
+    remote-address set ascii <client> local-address set
     1 minutes over set-timeout ;
 
+: https-tunnel? ( request -- ? )
+    [ proxy-url>> ] [ url>> protocol>> "https" = ] bi and ;
+
+: ?copy-proxy-basic-auth ( dst-request src-request -- dst-request )
+    proxy-url>> [ username>> ] [ password>> ] bi 2dup and
+    [ set-proxy-basic-auth ] [ 2drop ] if ;
+
+: ?https-tunnel ( -- )
+    request get dup https-tunnel? [
+        <request> swap [ url>> >>url ] [ ?copy-proxy-basic-auth ] bi
+        f >>proxy-url "CONNECT" >>method write-request
+        read-response check-response drop opt-send-secure-handshake
+    ] [ drop ] if ;
+
+! Note: ipv4 addresses are interpreted as subdomains but "work"
+: no-proxy-match? ( host-path no-proxy-path -- ? )
+    dup first empty? [ [ rest ] bi@ ] when tail? ;
+
+: get-no-proxy-list ( -- list )
```

Comment:
What is weird ? There was a bug when no_proxy was the empty string or contained 2 consecutives "," (it would always match, ie disable the proxy). I've pushed bceadde which fixes that and add tests

After that's fixed, it's the same as curl:
```sh
# curl
$ http_proxy=http://localhost:3129 no_proxy="" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy=http://localhost:3129 no_proxy="com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="om" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy=http://localhost:3129 no_proxy=".com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="www.google.com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="www.ggle.com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy=http://localhost:3129 no_proxy="google.com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="oogle.com" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy=http://localhost:3129 no_proxy="www.google.com,www.google.fr" curl www.google.com &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="www.google.com,www.google.fr" curl www.google.fr &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy=http://localhost:3129 no_proxy="www.google.com" curl www.google.fr &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy=http://localhost:3129 no_proxy="216.58.211.99" curl 216.58.211.99 &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY

# factor
$ http_proxy="http://localhost:3129" no_proxy="" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy="http://localhost:3129" no_proxy="com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="om" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy="http://localhost:3129" no_proxy=".com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="www.google.com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="www.ggle.com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy="http://localhost:3129" no_proxy="google.com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="oogle.com" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy="http://localhost:3129" no_proxy="www.google.com,www.google.fr" ./factor -e='USING: http.client kernel ; "http://www.google.com" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="www.google.com,www.google.fr" ./factor -e='USING: http.client kernel ; "http://www.google.fr" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY
$ http_proxy="http://localhost:3129" no_proxy="www.google.com" ./factor -e='USING: http.client kernel ; "http://www.google.fr" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
PROXY
$ http_proxy="http://localhost:3129" no_proxy="216.58.211.99" ./factor -e='USING: http.client kernel ; "216.58.211.99" http-get 2drop' &> /dev/null ; if [[ $? == "0" ]]  ;then echo NOPROXY; else echo PROXY; fi
NOPROXY

```